### PR TITLE
acc: generating duplicates

### PR DESCRIPTION
**Pull request by jchavanton**
on missed / failed calls issue #1673

#### Pre-Submission Checklist

- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [x] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [x] Breaking change (not tested with all scenarios, review could confirm that this is not introducing regressions)

#### Checklist:

- [ ] PR should be backported to stable branches
- [x] Tested changes locally
- [x] Related to issue #1673 

#### Description
problem with confusion when using  flags (integer value vs bit position value)
proposed solution is using the flags helpers functions `isflagset`, etc. where possible.

### CODE REVIEW

File: src/modules/acc/acc_logic.c [hunk 1]
```diff
@@ -501,19 +501,18 @@ static inline void on_missed(struct cell *t, struct sip_msg *req,
 	 * forwarding attempt fails; we do not wish to
 	 * report on every attempt; so we clear the flags;
 	 */
-
 	if (is_log_mc_on(req)) {
 		env_set_text( ACC_MISSED, ACC_MISSED_LEN);
 		acc_log_request( req );
-		flags_to_reset |= log_missed_flag;
+		flags_to_reset |= 1 << log_missed_flag;
```
Comment (jchavanton) [inline src/modules/acc/acc_logic.c]: I wonder why this was not creating more problem in the past the flags variable where containing "bit position", this is one point that the reviewer can help to confirm. (I may triple check that this is always true)

**Comment by henningw** [comment]
Hello Julien, thank you for this request.

So did I understood it correctly, you want to unify the flag manipulation and also fix some bugs this way? It is probably better to not combine this two changes. If you change different modules, these changes should be also in individual commits, if they are not interdependent on each other.

I had a quick look to the patch.

-			if((type==1) && (msg->flags&(e->missed_flag))) {
+			if((type==1) && isflagset(msg, e->missed_flag) == 1) {

This should be identically, according to the definition of isflagset. I don't see any big advantage of one format to the other, or I am wrong?

One question about this change (it is represented several times):

-		_acc_diameter_engine.missed_flag = 1<<diameter_missed_flag;
+		_acc_diameter_engine.missed_flag = diameter_missed_flag;

This will not produce identical results, I think. The first one is a left shift of diameter_missed_flag length, and the second one an assignment. Example:

diameter_missed_flag                      = 00000000000000000000000000001000 (8)
_acc_diameter_engine.missed_flag = 00000000000000000000000010000000 (256)

What is your reasoning behind this change?

**Comment by jchavanton** [comment]
Hi Hennig, thanks for the review so far.

Q: "you want to unify the flag manipulation and also fix some bugs this way?"
A: Correct, to fix the bug we need to change inconsitant flag manipulation creating the bug 

Q: This should be identically, according to the definition of isflagset. I don't see any big advantage of one format to the other, or I am wrong?
A: They are not identical :

The first one is expecting an input we the integer contain the bit mask
`if((type==1) && (msg->flags&(e->missed_flag))) {` 
will with with the input of : 1,2,4,8,...,4294967296

The second one is expecting an input where integer contain the bit position
`if((type==1) && isflagset(msg, e->missed_flag) == 1) {`
will do the bit shifting and will work with the input : 1,2,3,..,32

```
int isflagset( struct sip_msg* msg, flag_t flag ) {
        return (msg->flags & (1<<flag)) ? 1 : -1; // << notice the bit shifting
}
```

`resetflag` and `setflag` are also doing the bit shifthing, seems like at one point someone got confused and introduce a mix which endedup doing resetflag on an alredy bit shifted input.
I guess this is also why acc_mod is not doing bit shifting on the module param while acc_diameter, acc_raduis and acc_json are (I am changing this in the patch)

**Comment by henningw** [comment]
Hello Julien, you are right - this is indeed confusing. So you are also saying (regarding my second remark) that the other acc modules are doing it wrong?

**Comment by jchavanton** [comment]
Just pushed a complete working solution, 

tested with `acc` 
```
modparam("acc", "log_missed_flag", 9)
modparam("acc", "db_missed_flag", 10)
```
and with `acc_json`
```
modparam("acc_json", "acc_flag", 2)
modparam("acc_json", "acc_missed_flag", 7)
```
`acc_radius` and `acc_diameter` using the exact same integration

**Comment by jchavanton** [comment]
I believe this fix is ready to be merged

**Comment by henningw** [comment]
Thank you Julien, I will merge this later today. I will need to do it manually, in order to split the commits in modules and core.

**Comment by jchavanton** [comment]
I see, should have known ... thanks

**Comment by henningw** [comment]
Merged manually into git master

**Pull request closed without merging**